### ISSUE TITLE
[AXE-CORE, SEMANTIC MARKUP]: Email inputs MUST be paired with labels

### ISSUE THREAD

**Issue by jenstrickland** [department-of-veterans-affairs/va.gov-team#9860]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

On the 1 of 3 Veteran information screen, there are text inputs for email addresses. At present, the label that says "Email address" is in a `<p>`, but must be in a `<label>` with a for attribute whose value matches the id of the input field. Additionally, h4 is not valid HTML as child of element label in this context.

## Point of Contact

**VFS Point of Contact:** Jennifer

## Acceptance Criteria

As a mouse user, I want to be able to select the label of an input to move focus into the input.

## Environment

* Operating System: all
* Browser: 
* Server destination: staging

## Steps to Recreate

1. Enter https://staging.va.gov/hearing-aid-batteries-and-accessories/veteran-information/addresses while logged in as user 1 in browser
2. Inspect the Email address code
3. Verify that the code appears as below in the "Current code"

### Current code
```diff
<div class="schemaform-field-template vads-u-margin-bottom--3 schemaform-first-field">
- <label id="root_email-label" class="schemaform-label" for="root_email">
    <h4>Email address</h4>
  </label>
  <p>We will send an order confirmation email with a tracking number to this email address.</p>
- <p class="vads-u-margin-bottom--1">Email address</p>
  <div class="schemaform-widget-wrapper">
-   <input type="email" id="root_email" name="root_email" maxlength="80" class="va-input-large" value="">
  </div>
  <div></div>
</div>
```

### Recommended code
```diff
<div class="schemaform-field-template vads-u-margin-bottom--3 schemaform-first-field">

  <h4>Email address</h4>

  <p>We will send an order confirmation email with a tracking number to this email address.</p>

+ <label id="root_email" class="schemaform-label" for="root_email">Email address</label>
+ <input id="root_email" type="email" name="root_email" maxlength="80" class="va-input-large"/>
</div>
```

## WCAG or Vendor Guidance (optional)

* [W3C Labeling Controls](https://www.w3.org/WAI/tutorials/forms/labels/)

* [MDN Basic Form Hints](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/forms/Basic_form_hints) - "Note that each `<input>` element has an id, and each `<label>` element has a for attribute, indicating the id of the associated `<input>`.

## Screenshots or Trace Logs

![Screen Shot 2020-06-04 at 5 12 54 PM](https://user-images.githubusercontent.com/57469/83811256-77449300-a687-11ea-9785-6840db1115d5.png)

**Comment by 1Copenut** [comment]
@jenstrickland I agree with the markup change you're proposing here. What do you think about making the `<h4>` say something like "Email address for tracking number"? Maybe that's too many words, but something to distinguish it from the input labels.

**Comment by jenstrickland** [comment]
@1Copenut I'm not sure, it depends on the purpose of the email address. I agree providing clarity in the h4 would be preferable. @rebeccawalsh and @rileyorr would likely know more.

**Comment by leahkeeler** [comment]
@rileyorr @rebeccawalsh Can you please provide input on what the label should be to @bradl3yC

**Comment by rebeccawalsh** [comment]
'Email for order confirmation'?

**Comment by leahkeeler** [comment]
@jenstrickland - This is a forms issue per @bradl3yC - I have added this to the list of forms issues @andreahewitt-odd has

**Comment by jenstrickland** [comment]
I just removed the 508-defect-1 ❗ Launchblocker labels.

**Comment by bkjohnson** [comment]
I've described the issue with a bit of detail in a comment (https://github.com/department-of-veterans-affairs/va.gov-team/issues/10095#issuecomment-675711544), pointing out that the issue is with how the form is configured.

Here in the form configuration you can see the [markup which isn't desired](https://github.com/department-of-veterans-affairs/vets-website/blob/15f0403d9b605c6108730249738024af74777991/src/applications/disability-benefits/2346/schemas/2346UI.js#L21) (the "Email address" text inside of a `<p>` tag):

```jsx
const emailUIDescription = (
  <>
    <p>
      We'll send an order confirmation email with a tracking number to this
      email address.
    </p>
    <p className="vads-u-margin-bottom--1">Email address</p>
  </>
);
```

Additionally, right above this line you can see ["Email address" inside an `<h4>`](https://github.com/department-of-veterans-affairs/vets-website/blob/15f0403d9b605c6108730249738024af74777991/src/applications/disability-benefits/2346/schemas/2346UI.js#L13)

```jsx
const emailUITitle = <h4>Email address</h4>;
```

This is what is passed into the `ui:title` property of the uiSchema.

If the `<h4>` isn't desired, then it should be removed from the form configuration and a plain string should be used in its place.

**Comment by meganhkelley** [comment]
Hey @leahkeeler @bradl3yC we're passing this one back to y'all, Brooks' comment above explains how to fix this in your app (not in forms library). Please reach out with any Qs!

**Comment by bradl3yC** [comment]
@jenstrickland @1Copenut I made the changes suggested by Brooks - they're merged and should be in staging in the next build for validation.  Axe does not show any errors or warnings from the nesting.

![Screen Shot 2020-08-20 at 2 10 04 PM](https://user-images.githubusercontent.com/6165421/90809001-dca33a00-e2ee-11ea-871b-55bed9bf9832.png)

**Comment by jenstrickland** [comment]
@1Copenut Would you weigh in, please?

Looking at the result, I have reservations about this structure. Putting the input's label in an h4 breaks best practices. It would be preferable if the h4 used @rebeccawalsh 's heading:

> 'Email for order confirmation' or 'Order confirmation email'?

Then, the email inputs would follow the pattern shown in the [Caregiver application](https://staging.va.gov/family-member-benefits/apply-for-caregiver-assistance-form-10-10cg/introduction), along with the email validation (proper email and validate in re-entry):

![Screen Shot 2020-08-24 at 2 16 52 PM](https://user-images.githubusercontent.com/57469/91081031-7a12ac80-e614-11ea-873e-8b32ff81ad68.png)

```
<div>
  <div class="schemaform-field-template">
    <label id="root_veteranEmail-label" class="schemaform-label" for="root_veteranEmail">
      Veteran's  email address
    </label>
    <div class="schemaform-widget-wrapper">
      <input type="email" id="root_veteranEmail" name="root_veteranEmail" maxlength="80" value="">
    </div>
    <div></div>
</div>
</div>
```

Right now, this is what I see when I enter invalid data.

![Screen Shot 2020-08-24 at 2 19 08 PM](https://user-images.githubusercontent.com/57469/91081269-da095300-e614-11ea-8faa-172690dbe588.png)

**Comment by bradl3yC** [comment]
@jenstrickland Im confused, I do not see a H4 in the markup you've posted.  We're no longer using any custom markup, this is 100% from the forms system.

**Comment by 1Copenut** [comment]
This heading, paragraph and labels should match the pattern @Clark-W-Griswold outlined in his comment here: https://github.com/department-of-veterans-affairs/va.gov-team/issues/10095#issuecomment-673644968

The heading "Email address" is associated with the first input in the Staging build, but it should be configured so that each input has its own visual and semantic label, that is not the Bitter typeface styled heading.

**Comment by bradl3yC** [comment]
@1Copenut his comment is that its a form configuration issue, which was addressed in a PR that removed all markup and simply uses the form config with no additional markup.  Is the ask now to put back some JSX to make it more semantic?

**Comment by meganhkelley** [comment]
Hi, piping in here since @bkjohnson made the recommendation on related ticket #10095. Can someone (perhaps @jenstrickland) please clarify the acceptance criteria for this issue? Currently, the ticket says: 

> As a mouse user, I want to be able to select the label of an input to move focus into the input.

Jennifer, if there are other requirements for what the end behavior needs to be (such as — it sounds like — the label or error needs to be formatted a certain way or say a certain thing?) could you please add those to the acceptance criteria? 

Definitely want us to solve this in a way that makes sense! Just don't want additional requirements that come up to get lost in the comment train. Thanks all! 

(Hoping this will also help answer @bradl3yC's Q) 👍

**Comment by jenstrickland** [comment]
@meganhkelley I'll modify the Acceptance Criteria. New:

> As a user with cognitive considerations, I expect to see a label and input pairing consistently styled as throughout the rest of the site, with the label just above the text/email/search input or to the right of a radio/checkbox input, so that I am clearly able to understand what entry is expected.

**Comment by meganhkelley** [comment]
@jenstrickland awesome, thank you so much! If we've got a screenshot of expected behavior (per your comment that it's styled this way on the rest of the site) that would be swell, but I'll let @bradl3yC clarify whether he needs clarification

**Comment by jenstrickland** [comment]
Here's a screenshot of the desired result, along with code. This deactivated `#root_vetEmail-label` styling.

![Screen Shot 2020-08-24 at 4 18 50 PM](https://user-images.githubusercontent.com/57469/91092108-86533580-e625-11ea-9639-acb51098993f.png)

**Comment by bradl3yC** [comment]
@jenstrickland so currently the title in the form/ui config of the widget is whats passed to the `<label>` and the description is placed underneath as a `<p>` tag.  No matter what I put in for a title, the forms system will wrap it in a label.  Maybe @bkjohnson has some insights on how to achieve this.

**Comment by bkjohnson** [comment]
Brad is right.  Right now things are rendered in the following order:

1. label
1. description
1. errors
1. input

We don't provide any way at the moment to change how these elements are ordered.

**Comment by bkjohnson** [comment]
Here is what we currently have - this screenshot is from the latest version of master.

![image](https://user-images.githubusercontent.com/2008881/91098817-032ae300-e617-11ea-88eb-9309f40ec2e4.png)


Based on your screenshot, is this what you are asking for?

- Add the `<h4>` back in and have it read "Email for order confirmation" instead of "Email address"
  - This looks to be part of the description
- Remove the [styling changes added in this PR](https://github.com/department-of-veterans-affairs/vets-website/pull/13951/files#diff-609cfc28f4c28dfb3715f91d660e2682)
- Change the component so that the "description" appears above the label

**Comment by jenstrickland** [comment]
@bkjohnson Is it possible to use the label+input pattern that the Caregiver team is using (comment above https://github.com/department-of-veterans-affairs/va.gov-team/issues/9860#issuecomment-679288576) for this set of label+input? Then, an h4 and a p could be provided above. 

The VAFS is a bit of a mystery to me, so I'm not sure what constraints are immovable.

**Comment by bkjohnson** [comment]
The configuration for the caregiver form that you're talking about doesn't appear to use the `ui:description` configuration option that the 2346 form does (this is where the `<h4>` used to be).

There isn't an immediately obvious way to do what I think you're asking regarding the `<h4>` and `<p>` tags - I think this would require a change to the forms system.

**Comment by meganhkelley** [comment]
FYI I added our labels just so that I can keep this on our radar since we've been solutioning/troubleshooting, I'm aware this may not require a forms library change.

**Comment by jenstrickland** [comment]
@1Copenut - looping you in to double-check my thinking. I think this should remain a ticket to modify the forms system, as this may be an issue for screen reader users and those with cognitive considerations.

1. Preferred choice would be to make a change to the forms system that would render this:

> Here's a screenshot of the desired result, along with code. This deactivated `#root_vetEmail-label` styling.
> 
> ![Screen Shot 2020-08-24 at 4 18 50 PM](https://user-images.githubusercontent.com/57469/91092108-86533580-e625-11ea-9639-acb51098993f.png)




<br/>
<br/>
2. BAM2 will ship with the following (since this is a 508-defect-2):

```
<!-- visually styled as an h4 --><label id="root_vetEmail-label" class="schemaform-label" for="root_vetEmail">Email address</label> 
<p>We'll send an order confirmation email with a tracking number to this email address.</p>
<div class="schemaform-widget-wrapper">
  <input type="email" id="root_vetEmail" name="root_vetEmail" maxlength="80" class="va-input-large" value="vets.gov.user+1@gmail.com">
</div>
```

**Comment by bradl3yC** [comment]
@jenstrickland is that not what we already have?  Are we still removing the styling added as Brooks inquired?

**Comment by jenstrickland** [comment]
@bradl3yC my proposal is that BAM2 ships with what you already have, and that the forms library is modified. I'm asking Trevor to confirm, since he's platform.

**Comment by bradl3yC** [comment]
@jenstrickland fantastic! Can we remove the BAM2 flags and close out the 508 epic for MDOT since this was the last remaining card?

**Comment by 1Copenut** [comment]
Please leave the ticket open just for now. Platform is going to meet to discuss this afternoon.

**Comment by bradl3yC** [comment]
@1Copenut absolutely, I just meant the over-arching 508 epic for MDOT 508 review which this ticket is linked to

**Comment by jenstrickland** [comment]
@bradl3yC I usually leave the 508 epic open until all the issues are addressed, even the platform ones (for example, [Public Websites audits](https://github.com/department-of-veterans-affairs/va.gov-team/issues?q=is%3Aissue+is%3Aopen+label%3Avsa-public-websites+label%3Aepic+label%3A508%2Faccessibility) with open issues). Does having this open epic count negatively against BAM2?

**Comment by bradl3yC** [comment]
@jenstrickland no but its not an accurate representation of what is being asked of our team to complete - all of the 508 issues that are specifically related to the MDOT that are required of BAM2 to complete have been completed.  Without context, those looking in on this project may think that we have outstanding work to complete on an already shipped feature, which is not the case.  It makes more sense to have a side-wide epic or a forms-system epic where these issues are tracked.

**Comment by jenstrickland** [comment]
@bradl3yC Each of the ticket titles are named so it is clear the open issues are platform issues. Individual tickets are not assigned to the team, and a note on the ticket indicates the team impacted. The accessibility audit is not a representation of the BAM2 team's work, but a 508 report of **the work shipped** which includes work beyond the BAM2 team. Our VSA teams collaborate with VSP and third-party tools when we ship.

Anyone who looks at this work and gives BAM2 any grief, direct them to me, please. **Y'all did great work! Seriously!**

**Comment by bradl3yC** [comment]
@jenstrickland I was going off of your comment on #9859 where you stated it is up to the team to decide.

**Comment by jenstrickland** [comment]
> @jenstrickland I was going off of your comment on #9859 where you stated it is up to the team to decide.

@bradl3yC — I said that because the team continues to ask and I don't understand the reason for differing from the documented process, although I also said I prefer to keep the epic open until all issues are resolved. At the bottom of the epic opening text it outlines a Definition of Done, and indicates the accessibility specialist will validate fixes and is responsible for closing the epic.  The process is also documented in the Shared Support Needs stand-up document, and has been reiterated often in our weekly meeting with PMs.

**Comment by bradl3yC** [comment]
@jenstrickland my apologies.  I was only going off of the comments visible to me related to this work.  I am not a PM and haven't been privy to those conversations :)

**Comment by meganhkelley** [comment]
Is there a specification for resolving this issue? I'm still unclear on whether this is something that FE Tools needs to address

**Comment by jenstrickland** [comment]
@meganhkelley Yes, please. The FE Tools team is being asked to modify the forms system component, as this can be an issue for screen reader users and those with cognitive considerations.

> 1. Preferred choice would be to make a change to the forms system that would render this:
> 
> > Here's a screenshot of the desired result, along with code. This deactivated `#root_vetEmail-label` styling.
> > ![Screen Shot 2020-08-24 at 4 18 50 PM](https://user-images.githubusercontent.com/57469/91092108-86533580-e625-11ea-9639-acb51098993f.png)
> 

Currently, making the `h4` a label is poor usability and an accessibility issue.

**Comment by bkjohnson** [comment]
> Currently, making the `h4` a label is poor usability and an accessibility issue.

At the moment there isn't an `<h4>` related to email addresses on the page.

Also, that markup is not an approved design right now, and the forms library doesn't support that structure as far as I am aware. Others can chime in here, but I believe we would need some kind of approval from the design system council in order to make that happen.  

Here is a PR that satisfies the acceptance criteria: https://github.com/department-of-veterans-affairs/vets-website/pull/14190

**Comment by jenstrickland** [comment]
What @1Copenut and I are requesting is that the forms library address this issue in order to support assistive tech needs, and ensure the label and input are next to each other, as in the recommended code. 

The label shouldn't be the h4 or anything so far away from the input (in this case the team had to ship an h4 with a paragraph in between, as well). The component should be architected to be next to each other as shown below (and I acknowledge the forms library wraps inputs in a `div`, that's fine):

```
<label for="email_a">Label for email input a</label>
<div class="schemaform-widget-wrapper">
  <input type="email" id="email_a" />
</div>
```

Trevor, is there anything I'm overlooking here?

**Comment by 1Copenut** [comment]
I agree with Jennifer's comment. We're looking to make sure the label is always next to its input and properly associated semantically. Reviewing the PR now.

**Comment by bkjohnson** [comment]
Following up from this [slack thread](https://dsva.slack.com/archives/C8E985R32/p1600898694011100), I'm going to pause work on this until the acceptance criteria have some more clarity and specification to them.

Linked from that thread, here is the current AC broken down into a list:

```
As a user with cognitive considerations, I expect to see a label and input pairing consistently styled as throughout the rest of the site, with the label just above the text/email/search input or to the right of a radio/checkbox input, so that I am clearly able to understand what entry is expected.

---

- [ ] All inputs have associated label elements
- [ ] Input labels are styled consistently
- [ ] Input labels appear immediately above the input OR to the right of a radio / checkbox
```

What isn't captured there is this additional markup that doesn't adhere to how the forms library renders fields:

> @meganhkelley Yes, please. The FE Tools team is being asked to modify the forms system component, as this can be an issue for screen reader users and those with cognitive considerations.
> 
> > 1. Preferred choice would be to make a change to the forms system that would render this:
> > 
> > > Here's a screenshot of the desired result, along with code. This deactivated `#root_vetEmail-label` styling.
> > > ![Screen Shot 2020-08-24 at 4 18 50 PM](https://user-images.githubusercontent.com/57469/91092108-86533580-e625-11ea-9639-acb51098993f.png)
> 
> Currently, making the `h4` a label is poor usability and an accessibility issue.

---

It is my understanding that PR https://github.com/department-of-veterans-affairs/vets-website/pull/14190 solves the problem according to the current acceptance criteria, but it isn't clear if the `<h4>` and the `<p>` at the top of the field are something that the app team wants, or that the accessibility experts want.

**Comment by cvalarida** [comment]
@jenstrickland Do those AC work for you? What would you change, remove or add?

**Comment by jenstrickland** [comment]
@cvalarida I think that is good. I wonder if we should think about making it bulletproof. I can imagine teams pushing the edges.

```
- [ ] If there is an additional info link paired with the label, 
      it must be placed between the label and the input
```

Is that helpful or distracting? 

Regarding Brooks' question about the h4 and p tags, I would expect that a team may want to have an h4 and a p to start a new section as they did in the example. Does the forms library making it all one component that I need to provide clarity for?

**Comment by bkjohnson** [comment]
Incorporating the additional info link sounds like it is out of scope for this particular issue, so I would save that for another issue possibly.

At the moment, the forms library doesn't have support for an `h4` and `p` to act as headers over a particular field. If that _is_ a common pattern, then I think that might be something for the design team to talk about. Also, I'll point out that in your screenshot of the desired markup, the `h4` and `p` tags are already nested within the field for "Email Address" - semantically they aren't encapsulating the "Re-enter email address" field. I'm not aware of any instances where there are "sub-groups" on a single page.If that is the goal, then I think that would definitely involve more work.

**Comment by jenstrickland** [comment]
To clarify, I'm not asking to incorporate the additional info link. I only raised that for awareness.

**Comment by meganhkelley** [comment]
@caw310 handing this off to the design system team as well :)
